python/src/py_nvbench.cpp

```diff
@@ -0,0 +1,723 @@
+/*
+ *  Copyright 2025 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+// clang-format off
+// Include Pybind11 headers first thing
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+// clang-format on
+
+#include <nvbench/nvbench.cuh>
+
+#include <cstdio>
+#include <cstdlib>
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace py = pybind11;
+
+namespace
+{
+
+struct PyObjectDeleter
+{
+  void operator()(py::object *p)
+  {
+    const bool initialized = Py_IsInitialized();
+
+#if PY_VERSION_HEX < 0x30d0000
+    const bool finalizing = _Py_IsFinalizing();
+#else
+    const bool finalizing = Py_IsFinalizing();
+#endif
+    const bool guard = initialized && !finalizing;
+
+    // deleter only call ~object if interpreter is active and
+    // not shutting down, let OS clean up resources after
+    // interpreter tear-down
+    if (guard)
+    {
+      delete p;
+    }
+  }
+};
+
+struct benchmark_wrapper_t
+{
+
+  benchmark_wrapper_t() = default;
+
+  explicit benchmark_wrapper_t(py::object o)
+      : m_fn{std::shared_ptr<py::object>(new py::object(std::move(o)), PyObjectDeleter{})}
+  {
+    if (!PyCallable_Check(m_fn->ptr()))
+    {
+      throw py::value_error("Argument must be a callable");
+    }
+  }
+
+  // Only copy constructor is used, delete copy-assign, and moves
+  benchmark_wrapper_t(const benchmark_wrapper_t &other)            = default;
+  benchmark_wrapper_t &operator=(const benchmark_wrapper_t &other) = delete;
+  benchmark_wrapper_t(benchmark_wrapper_t &&) noexcept             = delete;
+  benchmark_wrapper_t &operator=(benchmark_wrapper_t &&) noexcept  = delete;
+
+  void operator()(nvbench::state &state, nvbench::type_list<>)
+  {
+    if (!m_fn)
+    {
+      throw std::runtime_error("No function to execute");
+    }
+    // box as Python object, using reference semantics
+    auto arg = py::cast(std::ref(state), py::return_value_policy::reference);
+
+    // Execute Python callable
+    (*m_fn)(arg);
+  }
+
+private:
+  // Important to use shared pointer here rather than py::object directly,
+  // since copy constructor must be const (benchmark::do_clone is const member method)
+  std::shared_ptr<py::object> m_fn;
+};
+
+class nvbench_run_error : std::runtime_error
+{};
```

Comment:
suggestion: I don't think this exception can be thrown (no constructor, private inheritance etc). Maybe something along the lines of:

```suggestion
struct nvbench_run_error : std::runtime_error {
  using std::runtime_error::runtime_error;
};
```